src/diffusers/pipelines/pipeline_utils.py

```diff
@@ -832,6 +834,28 @@ def load_module(name, value):
             return True
 
         init_dict = {k: v for k, v in init_dict.items() if load_module(k, v)}
+        scheduler_types = None
+        if "scheduler" in expected_types:
+            scheduler_types = []
+            for scheduler_type in expected_types["scheduler"]:
+                if isinstance(scheduler_type, enum.EnumMeta):
+                    scheduler_types.extend(list(scheduler_type))
+                else:
+                    scheduler_types.extend([str(scheduler_type)])
+            scheduler_types = [str(scheduler).split(".")[-1].strip("'>") for scheduler in scheduler_types]
+
+        for key, (_, expected_class_name) in zip(init_dict.keys(), init_dict.values()):
```

Comment:
Since we already have the types extracted in `expected_types` can't we fetch them using the key and then check if the passed object is an instance of the type? If the expected type is an enum then we can just check if the passed obj class name exists in the keys?

Reply:
https://github.com/huggingface/diffusers/pull/10189#discussion_r1880030678

Reply:
I think it might be better to make this check more agnostic to the component names. 

We have a few pipelines with Union types on non-scheduler components (mostly AnimateDiff). So this snippet would fail even though it's valid, because init_dict is based on the model_index.json which doesn't support multiple types.  

```python
from diffusers import (
	AnimateDiffPipeline,
	UNetMotionModel,
)

unet = UNetMotionModel()
pipe = AnimateDiffPipeline.from_pretrained(
	"hf-internal-testing/tiny-sd-pipe", unet=unet
)
```

Enforcing scheduler types might be a breaking change cc: @yiyixuxu . e.g. Using DDIM with Kandinsky is currently valid, but with this change any downstream code doing this it would break. It would be good to enforce on the Flow based schedulers though (perhaps via a new Enum)

I would try 

```python
        for key, (_, expected_class_name) in zip(init_dict.keys(), init_dict.values()):
            if key not in passed_class_obj:
                continue

            class_obj = passed_class_obj[key]
            _expected_class_types = []
            for expected_type in expected_types[key]:
                if isinstance(expected_type, enum.EnumMeta):
                    _expected_class_types.extend(expected_type.__members__.keys())
                else:
                    _expected_class_types.append(expected_type.__name__)

            _is_valid_type = class_obj.__class__.__name__ in _expected_class_types
            if isinstance(class_obj, SchedulerMixin) and not _is_valid_type:
                # Handle case where scheduler is still valid 
                # raise if scheduler is meant to be a Flow based scheduler?
            elif not _is_valid_type:
                raise ValueError(f"Expected types for {key}: {_expected_class_types}, got {class_obj.__class__.__name__}.")


```

Reply:
Added this for scheduler

```

Reply:
I think we don't need a value error here, a warning is enough, no?

Reply:
A warning should be sufficient, it's mainly for the situation here https://github.com/huggingface/diffusers/issues/10093#issue-2713961860 where the wrong text encoder is given because the resulting error is uninformative.

Reply:
let's do a warning then:)

Reply:
Just chiming here a bit to share a perspective as a user (not a strong opinion). Related to https://github.com/huggingface/diffusers/pull/10189#discussion_r1884592432. 

Here

https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_loading_utils.py#L287-L290

if there's an unexpected module passed we raise a value error. I think the check is almost along similar lines -- users are passing assigning components that are unexpected / incompatible. We probably cannot predict the consequences of allowing the loading without raising any errors but if we raise an error, users would know what to do to fix the in correct behaviour.